File: sdks/python/src/autonoma/check.py
"""Dry-run a scenario through the in-process handler.

``check_scenario`` lets a caller validate a scenario without standing up a
real HTTP server: it builds a :class:`HandlerConfig` with the single
scenario, invokes ``handle_request`` for the ``up`` and ``down`` phases,
and reports any errors.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from typing import Any

from autonoma.handler import HandlerRequest, handle_request
from autonoma.hmac_util import sign_body
from autonoma.types import HandlerConfig, ScenarioDefinition


@dataclass
class CheckError:
    phase: str
    message: str
    fix: str = ""


@dataclass
class CheckResult:
    valid: bool
    phase: str
    errors: list[CheckError] = field(default_factory=list)
    timing: dict[str, int] | None = None


async def check_scenario(
    scenario: ScenarioDefinition,
    options: dict[str, Any] | None = None,
) -> CheckResult:
    """Run a full ``up -> down`` cycle and return structured errors."""
    options = options or {}
    shared_secret = options.get("sharedSecret", "autonoma-check-shared")
    signing_secret = options.get("signingSecret", "autonoma-check-signing")
    test_run_id = options.get("testRunId", f"check-{scenario.name}")

    config = HandlerConfig(
        shared_secret=shared_secret,
        signing_secret=signing_secret,
        scenarios=[scenario],
    )

    up_body = json.dumps(
        {
            "action": "up",
            "scenario": {"name": scenario.name},
            "testRunId": test_run_id,
        }
    )
    up_req = HandlerRequest(
        body=up_body, headers={"x-signature": sign_body(up_body, shared_secret)}
    )

    t0 = time.monotonic()
    up_res = await handle_request(config, up_req)
    up_ms = round((time.monotonic() - t0) * 1000)

    if up_res.status != 200:
        message = up_res.body.get("error", "Unknown error")
        return CheckResult(
            valid=False,
            phase="up",
            errors=[CheckError(phase="up", message=message)],
            timing={"upMs": up_ms, "downMs": 0},
        )

    teardown_token = up_res.body.get("teardownToken", "")
    down_body = json.dumps(
        {
            "action": "down",
            "teardownToken": teardown_token,
        }
    )
    down_req = HandlerRequest(
        body=down_body, headers={"x-signature": sign_body(down_body, shared_secret)}
    )

    t1 = time.monotonic()
    down_res = await handle_request(config, down_req)
    down_ms = round((time.monotonic() - t1) * 1000)

    if down_res.status != 200:
        message = down_res.body.get("error", "Unknown error")
        return CheckResult(
            valid=False,
            phase="down",
            errors=[CheckError(phase="down", message=message)],
            timing={"upMs": up_ms, "downMs": down_ms},
        )

    return CheckResult(
        valid=True,
        phase="ok",
        errors=[],
        timing={"upMs": up_ms, "downMs": down_ms},
    )

File: sdks/elixir/lib/autonoma/check.ex
defmodule Autonoma.Check do
  @moduledoc "Dry-run a scenario against factories (up -> down cycle)."

  alias Autonoma.{HMAC, Handler}

  defmodule CheckError do
    @moduledoc false
    defstruct [:phase, :message, :fix]
  end

  defmodule CheckResult do
    @moduledoc false
    defstruct valid: false, phase: "ok", errors: [], timing: nil
  end

  def check_scenario(scenario, opts \\ []) do
    shared_secret = Keyword.get(opts, :shared_secret, "autonoma-check-shared")
    signing_secret = Keyword.get(opts, :signing_secret, "autonoma-check-signing")

    config = %{
      scope_field: Keyword.get(opts, :scope_field, "organizationId"),
      shared_secret: shared_secret,
      signing_secret: signing_secret,
      allow_production: Keyword.get(opts, :allow_production, true),
      sdk: Keyword.get(opts, :sdk, %{}),
      factories: Keyword.get(opts, :factories, %{}),
      auth:
        Keyword.get(opts, :auth, fn _user, _ctx ->
          %{"headers" => %{"Authorization" => "Bearer check-token"}}
        end)
    }

    create = Map.get(scenario, "create", Map.get(scenario, :create, %{}))

    # Up
    up_body = Jason.encode!(%{"action" => "up", "create" => create})
    up_sig = HMAC.sign_body(up_body, shared_secret)

    t0 = System.monotonic_time(:millisecond)
    up_res = Handler.handle(config, %{body: up_body, headers: %{"x-signature" => up_sig}})
    up_ms = System.monotonic_time(:millisecond) - t0

    if up_res.status != 200 do
      error_msg = Map.get(up_res.body, "error", "Unknown error")

      %CheckResult{
        valid: false,
        phase: "up",
        errors: [%CheckError{phase: "up", message: error_msg, fix: suggest_fix(error_msg)}],
        timing: %{up_ms: up_ms, down_ms: 0}
      }
    else
      refs_token = Map.get(up_res.body, "refsToken")

      # Down
      down_body = Jason.encode!(%{"action" => "down", "refsToken" => refs_token})
      down_sig = HMAC.sign_body(down_body, shared_secret)

      t1 = System.monotonic_time(:millisecond)
      down_res = Handler.handle(config, %{body: down_body, headers: %{"x-signature" => down_sig}})
      down_ms = System.monotonic_time(:millisecond) - t1

      if down_res.status != 200 do
        error_msg = Map.get(down_res.body, "error", "Unknown error")

        %CheckResult{
          valid: false,
          phase: "down",
          errors: [%CheckError{phase: "down", message: error_msg}],
          timing: %{up_ms: up_ms, down_ms: down_ms}
        }
      else
        %CheckResult{
          valid: true,
          phase: "ok",
          errors: [],
          timing: %{up_ms: up_ms, down_ms: down_ms}
        }
      end
    end
  end

  def check_all_scenarios(scenarios, opts \\ []) do
    Enum.map(scenarios, fn scenario ->
      check_scenario(scenario, opts)
    end)
  end

  defp suggest_fix(msg) do
    cond do
      String.contains?(msg, "missing required fields") ->
        "A required field is missing. Add it to the entity with a value."

      String.contains?(msg, "no factory registered") ->
        "Register a factory for the referenced model."

      String.contains?(msg, "cycle detected") ->
        "Break the circular dependency between models."

      true ->
        nil
    end
  end
end
